managers/BaseManager.js

```diff
@@ -7,54 +7,134 @@ const config = require('../common/config');
 const eventmesh = require('../eventmesh');
 const CONST = require('../common/constants');
 const NotImplementedBySubclass = errors.NotImplementedBySubclass;
+const Conflict = errors.Conflict;
 
 class BaseManager {
 
-  registerWatcher() {
-    throw new NotImplementedBySubclass('registerWatcher');
+  registerWatcher(resourceGroup, resourceType, queryString) {
+    logger.debug(`Registering watcher for resourceGroup ${resourceGroup} of type ${resourceType} with queryString as ${queryString}`);
+    return eventmesh.apiServerClient.registerWatcher(resourceGroup, resourceType, this.handleResource, queryString)
+      .then(stream => {
+        logger.info(`Successfully set watcher with query string:`, queryString);
+        return setTimeout(() => {
+          try {
+            logger.info(`Refreshing stream after ${CONST.APISERVER.WATCHER_REFRESH_INTERVAL}`);
+            stream.abort();
+            return this.registerWatcher();
```

Comment:
If you use `Promise.delay`, synchronously handling errors here is not required and you can use `Bluebird catch` for any internal errors

Reply:
If my understanding is correct it will mean that call will never get resolved.

Reply:
You can get around this by using a catch in the promise chain. Basic code flow as below:
```
Promise.delay(CONST.APISERVER.WATCHER_REFRESH_INTERVAL)
  .then(()=>{
      logger.info(`Refreshing stream after ${CONST.APISERVER.WATCHER_REFRESH_INTERVAL}`); 
      stream.abort();
      return this.registerWatcher();
    })
    .catch(err=> logger.error('Failed registering watcher with error', e)
```

Reply:
One thing to ponder on is that we would like to co-locate similar Managers into a single BOSH jobs. For example, Backup manager and Restore Manager together will go to a job. In that case, there would be single file which calls both the managers, or instantiate both. Will that be possible if the registerwatcher is a recursive call?